python/taichi/math/mathimpl.py

```diff
@@ -469,71 +469,111 @@ def mod(x, y):
 
 
 @ti.func
-def rotate2d(p, ang):
-    """Rotates a 2d vector by a given angle in counter-clockwise.
+def translate(dx, dy, dz):
+    """Constructs a translation Matrix with shape (4, 4).
 
     Args:
-        p (:class:`~taichi.math.vec2`): The 2d vector to rotate.
-        ang (float): Angle of rotation, in radians.
+        dx (float): delta x.
+        dy (float): delta y.
+        dz (float): delta z.
 
     Returns:
-        :class:`~taichi.math.vec2`: The vector after rotation.
+        :class:`~taichi.math.mat4`: translation matrix.
 
     Example::
 
-        >>> from taichi.math import *
-        >>> @ti.kernel
-        >>> def test():
-        >>>     v = vec2(1, 0)
-        >>>     print(rotate2d(v, radians(30)))
-        [0.866025, 0.500000]
+        >>> import math
+        >>> ti.Matrix.translate(1, 2, 3)
+        [[ 1 0 0 1]
+         [ 0 1 0 2]
+         [ 0 0 1 3]
+         [ 0 0 0 1]]
     """
-    ca, sa = ti.cos(ang), ti.sin(ang)
-    x, y = p
-    return vec2(x * ca - p.y * sa, x * sa + y * ca)
+    return mat4([[1., 0., 0., dx], [0., 1., 0., dy], [0., 0., 1., dz],
+                 [0., 0., 0., 1.]])
 
 
 @ti.func
-def rotate3d(p, axis, ang):
-    """Rotates a vector in 3d space, given an axis and angle of rotation.
-
-    The vector `axis` should be a unit vector.
-
-    See "https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula"
+def scale(sx, sy, sz):
+    """Constructs a scale Matrix with shape (4, 4).
 
     Args:
-        p (:class:`~taichi.math.vec3`): The 3d vector to rotate.
-        axis (:class:`~taichi.math.vec3`): Axis of rotation.
-        ang (float): Angle of rotation, in radians.
+        sx (float): scale x.
+        sy (float): scale y.
+        sz (float): scale z.
+
+    Returns:
+        :class:`~taichi.math.mat4`: scale matrix.
 
     Example::
 
-        >>> from taichi.math import *
-        >>> @ti.kernel
-        >>> def test():
-        >>>     v = vec3(1, 0, 0)
-        >>>     axis = normalize(vec3(1, 1, 1))
-        >>>     print(rotate3d(v, axis, radians(30)))
-        [0.910684, 0.333333, -0.244017]
+        >>> import math
+        >>> ti.Matrix.scale(1, 2, 3)
+        [[ 1 0 0 0]
+         [ 0 2 0 0]
+         [ 0 0 3 0]
+         [ 0 0 0 1]]
+    """
+    return mat4([[sx, 0., 0., 0.], [0., sy, 0., 0.], [0., 0., sz, 0.],
+                 [0., 0., 0., 1.]])
+
+
+@ti.func
+def rot3d_by_axis(ang, axis):
+    """rotate the matrix by an angle with the vector as the rotation axis
```

Comment:
```suggestion
def rot3d_by_axis(axis, ang):
    """Returns the 3x3 matrix representation of the rotation with given axis `axis` and angle `ang`. The vector `axis` must be normalized.
```

Reply:
Quite confusing the name `rot3d_xxx` returns a 4x4 matrix here.

Reply:
But that's our final name, maybe here, the meaning of rot3d_xxx is rotation in 3D space, but the 4x4 matrix means a affine transformation

Reply:
Shall we hide  "3d" in the name and just call it `rot_by_axis`?

Reply:
But there are still `rotation2d`, `rotation3d`, which need us to change their names if so.